src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    class Node<T> {
+        private T data;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev,T data,Node<T> next) {
+            this.data = data;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node head = null;
+    private Node tail = null;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null,value,null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            newNode.next = tail.next;
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node newNode = new Node(null, value, null);
+        if (index == size) {
+            add(value);
+            size--;
+        }
+        if (index == 0) {
+            getNodeByIndex(index).prev = newNode;
+            newNode.next = getNodeByIndex(index);
+            head = newNode;
+        }
+        if (index > 0 && index < size) {
+            newNode.prev = getNodeByIndex(index - 1);
+            newNode.next = getNodeByIndex(index);
+            getNodeByIndex(index).prev = newNode;
+            getNodeByIndex(index - 1).next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        return (T) getNodeByIndex(index).data;
```

Comment:
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-create-repeating-code
This is code repeating. You always call **indexValidation** before you call **getNodeByIndex**
![image](https://user-images.githubusercontent.com/20326858/219942637-8fc86882-3484-49ec-85ec-e2f917544766.png)
![image](https://user-images.githubusercontent.com/20326858/219942667-a5d20fe5-0694-4cf5-b077-92c7856781e0.png)